distributed/rpc/rl/distributed_model.py

```diff
@@ -0,0 +1,287 @@
+                )
+            )
+
+        # wait until all obervers have finished this episode
+        for fut in futs:
+            fut.wait()
+
+    def finish_episode(self, ctx_id):
+        r"""
+        This function is mostly borrowed from the Reinforcement Learning example.
+        See https://github.com/pytorch/examples/tree/master/reinforcement_learning
+        The main difference is that it joins all probs and rewards from
+        different observers into one list, and uses the minimum observer rewards
+        as the reward of the current episode.
+        """
+
+        # joins probs and rewards from different observers into lists
+        R, probs, rewards = 0, [], []
+        for ob_id in self.rewards:
+            probs.extend(self.saved_log_probs[ob_id])
+            rewards.extend(self.rewards[ob_id])
+
+        # use the minimum observer reward to calculate the running reward
+        min_reward = min([sum(self.rewards[ob_id]) for ob_id in self.rewards])
+        self.running_reward = 0.05 * min_reward + (1 - 0.05) * self.running_reward
+
+        # clear saved probs and rewards
+        for ob_id in self.rewards:
+            self.rewards[ob_id] = []
+            self.saved_log_probs[ob_id] = []
+
+        policy_loss, returns = [], []
+        for r in rewards[::-1]:
+            R = r + args.gamma * R
+            returns.insert(0, R)
+        returns = torch.tensor(returns)
+        returns = (returns - returns.mean()) / (returns.std() + self.eps)
+        for log_prob, R in zip(probs, returns):
+            policy_loss.append(-log_prob * R)
+        policy_loss = torch.cat(policy_loss).sum()
+        dist_autograd.backward([policy_loss])
+
+        # collect gradients from observers
+        futs = []
+        # should be able to optimize this loop with a gather
+        for ob_rref in self.ob_rrefs:
+            futs.append(_async_remote_method(Observer.get_gradients, ob_rref, ctx_id))
+
+        grads = [fut.wait() for fut in futs]
+        grads = [*zip(*grads)]
+        grads = [sum(grad) for grad in grads]
+
+        # we can add a mode to directly accumulate gradients on the model
+        # instead of putting it into a separate ctx
+        # set grads for observer model
+        ob_params = list(self.ob_policy.parameters())
+        for g, p in zip(grads, ob_params):
+            p.grad = g
+
+        # set grads for agent model
+        ctx_grads = dist_autograd.get_gradients(ctx_id)
+        for p in self.policy.parameters():
+            p.grad = ctx_grads[p]
+
+        self.optimizer.step()
+
+        ob_params = [p.data.detach() for p in self.ob_policy.parameters()]
+        # should be able to optimize this loop with a scatter
+        futs = []
+        for ob_rref in self.ob_rrefs:
+            futs.append(_async_remote_method(Observer.update_model, ob_rref, ob_params))
+        for fut in futs:
+            fut.wait()
+
+
```

Comment:
I believe the above few steps can be replaced by c10d gather/scatter, and wrapped into your own version of distributed optimizer. In that way, it will be more efficient and look better.